### PR TITLE
Separate out tests of optional dependencies, add new unit test job

### LINKED ISSUE #46: Separate out tests of optional dependencies
# Background

The origin of this issue is from #30 , and subsequent PR #45 to resolve that issue.

> In the case of https://github.com/dsgibbons/shap/issues/30 , it's problematic because: shap should support numpy 1.24 (and indeed does, we merged PRs to remove deprecation warnings for that very purpose).
> Yet because of tensorflow (an optional dependency), we are forced to test in an environment with numpy 1.23.5. I.e., we might miss bugs that only appear in the higher package versions.

Put very simply:

- In our `run_tests` Actions pipeline, we are 1) pip installing ALL test dependencies (including optional dependencies like tensorflow, pytorch, catboost, etc.), then 2) running `pytest` to run all tests.
- Because of 1), it translates to the fact that we are only testing in an environment which fulfils all of the optional requirements. This is not representative of how end users might use `shap`.
- The particular instance where this becomes an issue is described in #30 and the excerpt above.

# Solutions

Would like to discuss some potential solutions. Here's one idea, just throwing it out there:

- We will split out the testing of optional dependencies' code & shap functionality into different steps within the same `run_tests` Actions pipeline.
- I'm proposing to keep the tests within the same Stage so that we get to reuse as much installed dependencies as we can. Alternatively, if we have implemented caching of package installations, then it's a non-issue.
- Then, we could do something like: `pip install -e '.[test-core]'`, `python -m pytest -m core` to run all the core tests. Of course, the pre-requisite here is that we decorate all test functions relying only on core-functionality with `@pytest.mark.core`.
- Once that passes, then we go on with `pip install -e '.[test-torch]', `python -m pytest -m torch` to run all pytorch related tests. Etc.etc.

Obviously this comes with a lot of downsides, 1) it complicates the testing setup and procedure in the local environment -- we would probably need to rely on scripts (either with pdm/poetry , Makefiles, or the good-ol' bash scripts) then, 2) potentially lengthens the testing CI runtime, and 3) all or most of our test functions will need to be decorated with pytest markers, which is a huge change to make.

So I'm not thrilled with this idea, but it does make our testing a bit more rigorous. Hence, I would say this issue is a low priority, but I'm really curious how other projects are dealing with this similar problem..

### PR DESCRIPTION
**Pull request by connortann**
Closes #46

Defines two dependency groups, `test-core` and `test-extras`. The group `tests` remains as the union of both.

Adds an extra CI job to run tests with core dependencies only.

### CODE REVIEW

File: tests/explainers/test_linear.py
Comment (thatlittleboy) [inline tests/explainers/test_linear.py:40]: The point of this test is to check that an error is raised if an invalid masker is passed in, not really about the Image masker.

So instead of skipping because `cv2` doesn't exist, let's just change the masker to something other than the accepted ones (which I think is Partition, Exact, and Independent?) which doesn't depend on any extra dependencies, if possible?

**Review by thatlittleboy — COMMENTED** [review]
Thanks for this, but I think this one needs some further deliberation. (Also don't really want to merge this in now, since it might slow down our PRs).

I'll come back to this later on.

**Comment by connortann** [comment]
> Thanks for this, but I think this one needs some further deliberation. 

Sure, no problem. Thanks for the review comments!

> (Also don't really want to merge this in now, since it might slow down our PRs).

May I ask, what in these changes do you think may slow down PRs? Are you concerned about the total time for GH Checks to complete, or something else?

**Comment by thatlittleboy** [comment]
Initially I was worried about the total time, but I see now it's not going to be an issue.

File: tests/explainers/test_linear.py
Comment (dsgibbons) [reply tests/explainers/test_linear.py:40]: I can take a quick look at this. I can raise a new issue and address this in a new PR.

Comment (dsgibbons) [reply tests/explainers/test_linear.py:40]: See #64.

Comment (dsgibbons) [reply tests/explainers/test_linear.py:40]: Fixed by #67

**Review by dsgibbons — CHANGES_REQUESTED** [review]
I think this is a pragmatic PR that addresses the core concerns of #30 and #46. Long-term we may need a more comprehensive solution, perhaps using something akin to the decorator-based approach proposed by @thatlittleboy. For now, however, I think this is a good approach that reuses the existing `pytest.importorskip` paradigm that is pervasive throughout the test suite.

I approve this PR in principle unless @thatlittleboy has any strong objections. Before I approve, let's just remove the ``pytest.importorskip("cv2")` that has been fixed by #67.

**Review by dsgibbons — APPROVED** [review]
Approved, just pending the nod from @thatlittleboy

**Pull request merged**
